crates/re_space_view_spatial/src/visualizers/utilities/labels.rs

```diff
@@ -0,0 +1,88 @@
+use re_entity_db::InstancePathHash;
+use re_log_types::{EntityPath, Instance};
+use re_viewer_context::ResolvedAnnotationInfos;
+
+use super::entity_iterator::clamped;
+
+#[derive(Clone)]
+pub enum UiLabelTarget {
+    /// Labels a given rect (in scene coordinates)
+    Rect(egui::Rect),
+
+    /// Labels a given point (in scene coordinates)
+    Point2D(egui::Pos2),
+
+    /// A point in space.
+    Position3D(glam::Vec3),
+}
+
+#[derive(Clone)]
+pub struct UiLabel {
+    pub text: String,
+    pub color: egui::Color32,
+
+    /// The shape/position being labeled.
+    pub target: UiLabelTarget,
+
+    /// What is hovered if this label is hovered.
+    pub labeled_instance: InstancePathHash,
+}
+
+/// Maximum number of labels after which we stop displaying labels for that entity all together.
+///
+/// TODO(#4451): Hiding of labels should be configurable. This can be the heuristic for it.
+pub const MAX_NUM_LABELS_PER_ENTITY: usize = 30;
+
+pub fn process_labels_3d<'a>(
+    entity_path: &'a EntityPath,
+    positions: impl Iterator<Item = glam::Vec3> + 'a,
+    labels: &'a [re_types::components::Text],
+    colors: &'a [egui::Color32],
+    annotation_infos: &'a ResolvedAnnotationInfos,
+    world_from_obj: glam::Affine3A,
+) -> impl Iterator<Item = UiLabel> + 'a {
+    let labels = clamped(labels, annotation_infos.len());
+    let colors = clamped(colors, annotation_infos.len()); // Assumes colors is already populated with at least one color.
```

Comment:
Ideally such assumptions should not be hidden in a comment inside the function, but part of the docstring.

You should also consider adding an assert

```rs
debug_assert!(lables.is_empty() || !colors.is_empty(), "Cannot add labels without colors");
```

or a `log::warn_once` or something (anything but a silent, undocumented error 😬 )

Reply:
very good point!